Error:Execution failed for task ':sample:clean'

When I include PermissionsDispatcher code, my project fails to clean build with following error message: 

Error:Execution failed for task ':sample:clean'.
> Unable to delete file: C:\PermissionsDispatcher-master\PermissionsDispatcher-master\sample\build\intermediates\exploded-aar\PermissionsDispatcher-master\library\unspecified\jars\lint.jar

If I remove this PermissionDispatcher code including from build.gradle file then everything working fine.

> I think It has an issue with https://github.com/hotchemi/PermissionsDispatcher/blob/master/lint/src/main/java/permissions/dispatcher/CallOnRequestPermissionsResultDetector.java - last methof afterVisitClassDeclaration but don't know what you are doing over there and what is missing in my code so it is showing lint error. As you have worked on permission dispatcher, you have better idea on it what is missing.

Well, if that the case, then we should get an error when we ran `./gradlew lintDebug`

For instance, if I commented out `onRequestPermissionsResult` and run `./gradle lintDebug`

```java
@Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // NOTE: delegate the permission handling to generated method
        // CameraFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }
```

I get this error message

```
...
:app:lintDebug
Wrote HTML report to file:///Users/isogai_yoshinori/Downloads/Permission/app/build/outputs/lint-results-debug.html
Wrote XML report to file:///Users/isogai_yoshinori/Downloads/Permission/app/build/outputs/lint-results-debug.xml
Lint found 1 errors and 16 warnings
:app:lintDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:lintDebug'.
> Lint found errors in the project; aborting build.
  
  Fix the issues identified by lint, or add the following to your build script to proceed with errors:
  ...
  android {
      lintOptions {
          abortOnError false
      }
  }
  ...

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

And the error description is following

```xml
<issue id="NeedOnRequestPermissionsResult" severity="Error" message="Generated onRequestPermissionsResult method not called" category="Correctness" priority="5" summary="Call the "onRequestPermissionsResult" method of the generated PermissionsDispatcher class in the respective method of your Activity or Fragment" explanation="You are required to inform the generated PermissionsDispatcher class about the results of a permission request. In your class annotated with @RuntimePermissions, override the "onRequestPermissionsResult" method and call through to the generated PermissionsDispatcher method with the same name." errorLine1="@RuntimePermissions " errorLine2="^">
<location file="/Users/isogai_yoshinori/Downloads/Permission/app/src/main/java/com/taim/fragments/CameraFragment.java" line="27" column="1"/>
</issue>
```

which is what you get. 

I felt like that is an issue with Android Studio, not PermissionsDispatcher or lint, like I said at the beginning. So, unless, you provide me how to show a custom line error on the editor, I won't be able help you.

--- a/lint/src/main/java/permissions/dispatcher/CallOnRequestPermissionsResultDetector.java
+++ b/lint/src/main/java/permissions/dispatcher/CallOnRequestPermissionsResultDetector.java
@@ -1,0 +1,115 @@
+package permissions.dispatcher;
+
+import com.android.tools.lint.detector.api.Category;
+import com.android.tools.lint.detector.api.Detector;
+import com.android.tools.lint.detector.api.Implementation;
+import com.android.tools.lint.detector.api.Issue;
+import com.android.tools.lint.detector.api.JavaContext;
+import com.android.tools.lint.detector.api.Scope;
+import com.android.tools.lint.detector.api.Severity;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Set;
+
+import lombok.ast.Annotation;
+import lombok.ast.AstVisitor;
+import lombok.ast.ClassDeclaration;
+import lombok.ast.ForwardingAstVisitor;
+import lombok.ast.MethodDeclaration;
+import lombok.ast.MethodInvocation;
+import lombok.ast.VariableReference;
+
+public class CallOnRequestPermissionsResultDetector extends Detector implements Detector.JavaScanner {
+    public static final Issue ISSUE = Issue.create("NeedOnRequestPermissionsResult",
+            "Call the \"onRequestPermissionsResult\" method of the generated PermissionsDispatcher class in the respective method of your Activity or Fragment",
+            "You are required to inform the generated PermissionsDispatcher class about the results of a permission request. In your class annotated with @RuntimePermissions, override the \"onRequestPermissionsResult\" method and call through to the generated PermissionsDispatcher method with the same name.",
+            Category.CORRECTNESS,
+            5,
+            Severity.ERROR,
+            new Implementation(CallOnRequestPermissionsResultDetector.class, EnumSet.of(Scope.JAVA_FILE)));
+
+    static final Set<String> RUNTIME_PERMISSIONS_NAME = new HashSet<String>() {{
+        add("RuntimePermissions");
+        add("permissions.dispatcher.RuntimePermissions");
+    }};
+
+    @Override
+    public AstVisitor createJavaVisitor(JavaContext context) {
+        return new OnRequestPermissionsResultChecker(context);
+    }
+
+    private static class OnRequestPermissionsResultChecker extends ForwardingAstVisitor {
+        private final JavaContext context;
+        private boolean hasRuntimePermissionAnnotation;
+        private boolean hasOnRequestPermissionResultCall;
+        private String generatedClassName;
+        private ClassDeclaration classDeclaration;
+
+        private OnRequestPermissionsResultChecker(JavaContext context) {
+            this.context = context;
+        }
+
+        @Override
+        public boolean visitClassDeclaration(ClassDeclaration node) {
+            if (!context.isEnabled(ISSUE)) {
+                // stop executing lint for this class
+                return true;
+            }
+
+            classDeclaration = node;
+            generatedClassName = node.astName() + "PermissionsDispatcher";
+            return super.visitClassDeclaration(node);
+        }
+
+        @Override
+        public boolean visitAnnotation(Annotation node) {
+            String type = node.astAnnotationTypeReference().getTypeName();
+            if (!RUNTIME_PERMISSIONS_NAME.contains(type)) {
+                return super.visitAnnotation(node);
+            }
+
+            hasRuntimePermissionAnnotation = true;
+            return super.visitAnnotation(node);
+        }
+
+        @Override
+        public boolean visitMethodDeclaration(MethodDeclaration node) {
+            if (hasRuntimePermissionAnnotation && "public void onRequestPermissionsResult(int, java.lang.String[], int[])".equals(context.resolve(node).getSignature().trim())) {
+                return super.visitMethodDeclaration(node);
+            } else {
+                // ignore this node
+                return true;
+            }
+        }
+
+        @Override
+        public boolean visitMethodInvocation(MethodInvocation node) {
+            if (!hasRuntimePermissionAnnotation || hasOnRequestPermissionResultCall) {
+                return super.visitMethodInvocation(node);
+            }
+
+            if (!"onRequestPermissionsResult".equals(node.astName().astValue())) {
+                return super.visitMethodInvocation(node);
+            }
+
+            if (node.astOperand() instanceof VariableReference) {
+                VariableReference ref = (VariableReference) node.astOperand();
+                if (generatedClassName.equals(ref.astIdentifier().astValue())) {
+                    hasOnRequestPermissionResultCall = true;
+                }
+            }
+
+            return super.visitMethodInvocation(node);
+        }
+
+        @Override
+        public void afterVisitClassDeclaration(ClassDeclaration node) {
+            if (hasRuntimePermissionAnnotation && !hasOnRequestPermissionResultCall) {
+                context.report(ISSUE, context.getLocation(classDeclaration), "Generated onRequestPermissionsResult method not called");
+            }
+            super.afterVisitClassDeclaration(node);
+        }
+    }
+
+}

--- a/lint/src/main/java/permissions/dispatcher/CallOnRequestPermissionsResultDetector.java
+++ b/lint/src/main/java/permissions/dispatcher/CallOnRequestPermissionsResultDetector.java
@@ -20,14 +20,14 @@ import lombok.ast.MethodDeclaration;
 import lombok.ast.MethodInvocation;
 import lombok.ast.VariableReference;
 
-public class CallOnRequestPermissionsResultDetector extends Detector implements Detector.JavaScanner {
+public class CallOnRequestPermissionsResultDetector extends Detector implements Detector.ClassScanner {
     public static final Issue ISSUE = Issue.create("NeedOnRequestPermissionsResult",
             "Call the \"onRequestPermissionsResult\" method of the generated PermissionsDispatcher class in the respective method of your Activity or Fragment",
             "You are required to inform the generated PermissionsDispatcher class about the results of a permission request. In your class annotated with @RuntimePermissions, override the \"onRequestPermissionsResult\" method and call through to the generated PermissionsDispatcher method with the same name.",
             Category.CORRECTNESS,
             5,
             Severity.ERROR,
-            new Implementation(CallOnRequestPermissionsResultDetector.class, EnumSet.of(Scope.JAVA_FILE)));
+            new Implementation(CallOnRequestPermissionsResultDetector.class, EnumSet.of(Scope.CLASS_FILE)));
 
     static final Set<String> RUNTIME_PERMISSIONS_NAME = new HashSet<String>() {{
         add("RuntimePermissions");

--- a/lint/src/main/java/permissions/dispatcher/PermissionsDispatcherIssueRegistry.java
+++ b/lint/src/main/java/permissions/dispatcher/PermissionsDispatcherIssueRegistry.java
@@ -11,6 +11,7 @@ public class PermissionsDispatcherIssueRegistry extends IssueRegistry {
     @Override
     public List<Issue> getIssues() {
         return Arrays.asList(CallNeedsPermissionDetector.ISSUE,
-                NoCorrespondingNeedsPermissionDetector.ISSUE);
+                NoCorrespondingNeedsPermissionDetector.ISSUE,
+                CallOnRequestPermissionsResultDetector.ISSUE);
     }
 }

